### PR TITLE
Enable load_weights_only in checkpointing

### LINKED ISSUE #1993: Support weights-only checkpoint loading
## What Tinker supports

Tinker 0.23 intentionally exposes two checkpoint-loading APIs:

- [`TrainingClient.load_state(...)`](https://github.com/thinking-machines-lab/tinker/blob/main/src/tinker/lib/public_interfaces/training_client.py#L721-L752) sends `optimizer=False`: restore model weights and reset optimizer state.
- [`TrainingClient.load_state_with_optimizer(...)`](https://github.com/thinking-machines-lab/tinker/blob/main/src/tinker/lib/public_interfaces/training_client.py#L754-L779) sends `optimizer=True`: restore model weights and optimizer state.

Weights-only loading is needed for warm starts and stage transitions where parameters should transfer but stale Adam momentum and scheduler state should not.

## What SkyRL is missing

SkyRL currently loses this distinction:

1. [`LoadWeightsRequest`](https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl/tinker/api.py#L614-L618) does not expose the `optimizer` flag.
2. The [`load_weights` endpoint](https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl/tinker/api.py#L974-L997) creates `LoadWeightsInput` without it.
3. [`SkyRLTrainBackend.load_checkpoint`](https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl/backends/skyrl_train_backend.py#L1235-L1254) hardcodes `load_optimizer_states=True` and `load_lr_scheduler_states=True`.

As a result, `load_state()` and `load_state_with_optimizer()` both restore optimizer and scheduler state on the SkyRL backend.

## Feature Request

- Add the `load_optimizer` boolean to SkyRL load-weight request/input types.
- Thread it through the API and engine to `load_checkpoint`.
- Restore optimizer and scheduler state only when `load_optimizer=True`.
- Add a regression test that first creates nonzero Adam moments, then verifies:
  - `load_state()` resets `exp_avg`, `exp_avg_sq`, and scheduler state.
  - `load_state_with_optimizer()` preserves them.

### PR DESCRIPTION
**Pull request by j316chuck**
## Summary

- Preserve Tinker's `optimizer` flag through the load-weights API, engine, and JAX/SkyRL-Train backends.
- Restore optimizer and scheduler state only for `load_state_with_optimizer()`; weights-only loads leave the live optimizer untouched.

Fixes [#1993](https://github.com/NovaSky-AI/SkyRL/issues/1993).

## Testing

```bash
uv run --extra tinker --extra dev pytest tests/tinker/test_api_validation.py::test_load_weights_request_preserves_optimizer_choice tests/tinker/test_engine.py::test_process_load_weights_forwards_optimizer_choice
uv run --extra tinker --extra skyrl-train --extra ray --extra dev pytest tests/backends/skyrl_train/test_checkpoint_loading.py
```

All seven focused API, engine, and checkpoint-loading cases pass.

---

> [!NOTE]
> **Medium Risk**
> Checkpoint loading changes training state semantics; default `optimizer=true` preserves prior behavior, but weights-only loads could surprise callers who expect a full resume.
> 
> **Overview**
> Adds **weights-only checkpoint loading** by threading a `load_optimizer` flag from the Tinker API through the engine into all backends.
> 
> **API & engine:** `LoadWeightsRequest` exposes `optimizer` (default `true`). The engine passes `load_optimizer` on `LoadWeightsInput` into `backend.load_checkpoint`.
> 
> **Backends:** `AbstractBackend.load_checkpoint` now requires `load_optimizer`. JAX skips restoring optimizer state when it is `false`; SkyRL-Train only loads optimizer and LR scheduler states when requested; Ray/JAX wrappers forward the flag.
> 
> **Tests** cover API validation, engine forwarding, SkyRL-Train dispatch flags, and that weights-only loads leave the live optimizer untouched.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d7240b647723fcf8ab8d990175209aba1f17ca7b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: skyrl/tinker/api.py [hunk 1]
```diff
@@ -614,6 +614,7 @@ class SaveWeightsRequest(BaseModel):
 class LoadWeightsRequest(BaseModel):
     model_id: str
     path: str
+    optimizer: bool
```
Comment (j316chuck) [inline skyrl/tinker/api.py]:
```suggestion
    optimizer: bool = True
```

File: skyrl/tinker/types.py [hunk 2]
```diff
@@ -208,6 +208,7 @@ class SaveWeightsOutput(BaseModel):
 class LoadWeightsInput(BaseModel):
     source_model_id: str
     checkpoint_id: str
+    optimizer: bool
```
[the remark this answers is not in the document]
Comment (j316chuck) [reply skyrl/tinker/types.py]:
```suggestion
    optimizer: bool = True
```

File: skyrl/backends/jax.py [hunk 3]
```diff
@@ -965,11 +965,12 @@ def _insert_checkpoint_data(self, model_id: str, checkpoint_data: dict) -> None:
             )
 
         insert_adapter_state(adapter_index, self.lora_params, checkpoint_data["lora_weights"], rank)
-        insert_adapter_state(
-            adapter_index, nnx.state(self.optimizers[model_id]), checkpoint_data["optimizer_state"], rank
-        )
+        if optimizer:
+            insert_adapter_state(
+                adapter_index, nnx.state(self.optimizers[model_id]), checkpoint_data["optimizer_state"], rank
+            )
```
[the remark this answers is not in the document]
Comment (j316chuck) [reply skyrl/backends/jax.py]: this changes current implementation behavior so disregarding comment

File: skyrl/tinker/types.py [hunk 4]
```diff
@@ -208,6 +208,7 @@ class SaveWeightsOutput(BaseModel):
 class LoadWeightsInput(BaseModel):
     source_model_id: str
     checkpoint_id: str
+    load_optimizer: bool = True
```
Comment (kalectory) [inline skyrl/tinker/types.py:211]: clean

**Comment by avigyabb** [comment]
Hey Chuck, thanks so much for putting this up! One quick question- in https://github.com/NovaSky-AI/SkyRL/issues/1993 you mention that Tinker `load_state` resets the optimizer state, but in your PR I believe you leave the optimizer state as is? Do you think it would make more sense in your workloads to match the behavior of Tinker here? Lmk!

**Comment by j316chuck** [comment]
You're mixing two different things.

Tinker's `load_state` (`optimizer=False`) means **do not restore optimizer/scheduler from the checkpoint**. It does not mean "zero the live optimizer that is already on this client." The "optimizer resets" wording in Tinker's docs is about the common path: `create_training_client` / `create_training_client_from_state` builds a **new** client with a fresh optimizer, then loads weights only. Fresh live state + skip checkpoint optimizer == a reset. Same for our warm-start / stage-transition workloads.

This PR matches that:

- `load_state` → `load_optimizer=False` → skip checkpoint Adam/scheduler; leave the live optimizer object untouched
- `load_state_with_optimizer` → restore checkpoint optimizer/scheduler

Actively wiping live `exp_avg` / `exp_avg_sq` on weights-only load would be a different (and wrong) behavior for an already-running client: you'd throw away the current run's moments even though they did not come from the checkpoint. The checkpoint's stale moments are what we need to avoid, and skipping the restore does that.

The issue's "resets exp_avg" test description was imprecise for that reason. `tests/backends/skyrl_train/test_checkpoint_loading.py::test_weights_only_load_does_not_reset_live_optimizer` is the intended contract.

**Comment by avigyabb** [comment]
Ah, I misunderstood when you said 

- "[TrainingClient.load_state(...)](https://github.com/thinking-machines-lab/tinker/blob/main/src/tinker/lib/public_interfaces/training_client.py#L721-L752) sends optimizer=False: restore model weights and reset optimizer state." 

in the [issue description](https://github.com/NovaSky-AI/SkyRL/issues/1993).

Digging into this deeper, it seems like the [Tinker docs for `load_state(...)`](https://tinker-docs.thinkingmachines.ai/tinker/api-reference/trainingclient/?h=load_state#load_state) are a bit ambiguous as to whether or not they reset the optimizer 😭 . When I ran some test scripts, it seems that Tinker only allows `load_state` to be the **first** request of a client (causing a `400: LoadWeights is not permitted with seq_id N` error if not). This is a bit more disruptive, so wanted to check in with you first and see if this breaks any existing functionality on your end? Thanks so much for the discussion/clarification!

**Comment by j316chuck** [comment]
Closing in favor of #2071 which builds on top of this PR with one more Tinker api check thanks to @avigyabb 's observation.
